src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala

```diff
@@ -141,81 +148,81 @@ case class ExpiringMultiLevelCache[V](ttl: FiniteDuration,
   override def apply(key: String, genValue: () => Future[V])(implicit ec: ExecutionContext, scheduler: Scheduler): Future[V] = {
     // The local cache is always the first try. We'll only look the remote if the local value is missing or has expired
     val startTime = System.nanoTime()
-    val result = localCache.flatMap(_.get(key).map(_.asTry())) match {
+    val result: Future[V] = localCache.flatMap(_.get(key).map(_.asTry())) match {
       case Some(future) =>
         future.flatMap {
-          case Success(localValue) if !localValue.hasExpired(ttl, now) =>
+          case Success(localValue) if !localValue.hasExpired(ttl, now, ttlCachedErrors) =>
             // We have locally a good value, just return it
             reporter.onLocalCacheHit(key, elapsedTime(startTime))
             // But if we're paranoid, let's check if the local value is consistent with remote
             if (sanityLocalValueCheck)
-              remoteRW.map(remote => sanityLocalValueCheck(key, localValue, remote, genValue, startTime)).getOrElse(Future.successful(localValue.value))
+              remoteRW.map(remote => sanityLocalValueCheck(key, localValue, remote, genValue, startTime)).getOrElse(Future.successful(localValue.value.get))
             else
-              Future.successful(localValue.value)
+              Future.successful(localValue.value.get)
```

Comment:
You're not checking the localValue.value type.
If localValue.value is a failure, ".get" will throw an exception and you're wrapping it in a successful future.
When you're dealing with monad-like structures in scala, avoid the use of methods like "get". Prefer .map, .flatMap or pattern matching instead.